p2p/peer.go

```diff
@@ -214,97 +214,280 @@ type PeerUpdate struct {
 	Status PeerStatus
 }
 
-// peerStore manages information about peers. It is currently a bare-bones
-// in-memory store of peer addresses, and will be fleshed out later.
+// peerManager manages peer information, using a peerStore for underlying
+// storage. Its primary purpose is to determine which peers to connect to next,
+// make sure a peer only has a single active connection (either inbound or outbound),
+// and to avoid dialing the same peer in parallel goroutines.
 //
-// The main function of peerStore is currently to dispense peers to connect to
-// (via peerStore.Dispense), giving the caller exclusive "ownership" of that
-// peer until the peer is returned (via peerStore.Return). This is used to
-// schedule and synchronize peer dialing and accepting in the Router, e.g.
-// making sure we only have a single connection (in either direction) to peers.
-type peerStore struct {
-	mtx     sync.Mutex
-	peers   map[NodeID]*peerInfo
-	claimed map[NodeID]bool
+// For an outbound connection, the flow is as follows:
+// - DialNext: returns a peer address to dial, marking the peer as dialing.
+// - DialFailed: reports a dail failure, unmarking the peer as dialing.
+// - Dialed: successfully dialed, unmarking as dialing and marking as connected
+//   (or erroring if already connected).
+// - Ready: routing is up, broadcasts a PeerStatusUp peer update to subscribers.
+// - Disconnected: peer disconnects, unmarking as connected.
+//
+// For an inbound connection, the flow is as follows:
+// - Accepted: successfully accepted connection, marking as connected (or erroring
+//   if already connected).
+// - Ready: routing is up, broadcasts a PeerStatusUp peer update to subscribers.
+// - Disconnected: peer disconnects, unmarking as connected.
+//
+// We track dialing and connected states independently. This allows us to accept
+// an inbound connection while the router is also dialing an outbound
+// connection, which will cause the dialer to eventually error (when attempting
+// to mark the peer as connected). This also avoids race conditions where
+// multiple goroutines may end up dialing a peer if an incoming connection was
+// briefly accepted and disconnected while we were also dialing.
+type peerManager struct {
+	mtx           sync.Mutex
+	store         *peerStore
+	dialing       map[NodeID]bool
+	connected     map[NodeID]bool
+	subscriptions map[*PeerUpdatesCh]*PeerUpdatesCh // keyed by struct identity (address)
+}
+
+// newPeerManager creates a new peer manager.
+func newPeerManager(store *peerStore) *peerManager {
+	return &peerManager{
+		store:         store,
+		dialing:       map[NodeID]bool{},
+		connected:     map[NodeID]bool{},
+		subscriptions: map[*PeerUpdatesCh]*PeerUpdatesCh{},
+	}
+}
+
+// Add adds a peer to the manager, given as an address. If the peer already
```

Comment:
Looks like the peer address is (attempted to be) added regardless if the peer exists in the store.

Reply:
Yes, this is how we get new peers into the store in the first place. `peerManager.Add()` is used to add peers to the store, and the rest of Tendermint identifies peers by a peer address (e.g. in configuration files and via PEX).

Reply:
Yup, but I was referring to the last sentence in the godoc :)

Reply:
Ah. Yes, that's also accurate I believe -- if the peer exists, we try to add the address to it (in case it's a new address for an existing peer).